### ISSUE TITLE
wicked_pdf_image_tag Tag for ActiveStorage::Attachment not working

### ISSUE THREAD

**Issue by m-andreas** [mileszs/wicked_pdf#716]
wicked_pdf_image_tag Tag does not support Rails 5.2 ActiveStorage.

I created a workaround for people having this issue in the meantime:

    def wicked_image_active_storage_workaround( image )
      if image.is_a? ActiveStorage::Attachment
        save_path = Rails.root.join( "tmp", "#{image.id}.jpg")
        File.open(save_path, 'wb') do |file|
          file << image.blob.download
        end
        return save_path.to_s
      end
    end

In the view

    image_tag wicked_image_active_storage_workaround( image )

Keep in mind to clean up the tmp folder from time to time, or delete the image directly after using it.

Looking forward to a clean solution of this!

**Comment by nlinn** [comment]
I ran into the same issue, and simply skipped to tag helper like this:

`<img src="<%= file.service_url %>">`

or if you need resizing, like this:

`<img src="<%= file.variant(resize: "590").processed.service_url %>">`

**Comment by unixmonkey** [comment]
These examples are great. I'd love to see a PR that makes WickedPDF work with ActiveStorage better.

I think the first example would be almost good enough (but perhaps named something like `wicked_active_storage_asset`, and maybe using `WickedPdfTempfile`).

Do you know if this same technique would work for an AS::Blob or non-attachment, or are they all attachments at that point?

**Comment by rubyistdotjs** [comment]
None of the above worked for me, most likely due to the fact that I generate the PDF in a background job. Here's what I ended up with :

```ruby
def wicked_blob_path(file)
  ActiveStorage::Blob.service.send(:path_for, file.blob.key)
end
```

```slim
= image_tag(wicked_blob_path(user.avatar), alt: '_')
```

**Comment by cseelus** [comment]
Although using `service_url` works for my apps when I tested Active Storage (its not ideal though, see [Rails API](https://api.rubyonrails.org/v5.2.0/classes/ActiveStorage/Variant.html#method-i-service_url)), it would be nice if we could just keep using `image_tag` without any additional helpers like in the first example, which would only work with template files specifically made for Wicked PDF.

After all what makes Wicked PDF so great is that one can use the same template files for serving Webbrowsers and creating PDF files.

I use Wicked PDF for around 3 years now on my background workers (thanks @unixmonkey 💯), and never had any problems with my images (I use Shrine instead of Active Storage).

**Comment by macfanatic** [comment]
Here is the solution we ended up using, allowing us to use Disk storage in development and S3 on production:

```ruby
module PDFHelper
  def wicked_blob_path(active_storage_attachment)
    service = active_storage_attachment.service

    case service
    when ActiveStorage::Service::DiskService
      service.path_for(active_storage_attachment.blob.key)
    when ActiveStorage::Service::S3Service
      active_storage_attachment.service_url
    else
      raise "Unsupported ActiveStorage service for WickedPDF integration: #{service.name}"
    end
  end
end
```

In our view files we used this helper as normal:

```haml
%section
  = image_tag(wicked_blob_path(organization.logo))
```

**Comment by katatsu12** [comment]
As an option:
```
= image_tag polymorphic_url image
```

**Comment by igorgesso** [comment]
> As an option for `ActiveStorage::Service::DiskService`:
> 
> ```
> = image_tag polymorphic_url image
> ```

I couldn't get polymorphic to work, i'm using rails api and trying to get the error msg "`polymorphic_url' for #<ActionView::Base".

When trying .service_url the error too, it looks like it's been discontinued.

Is there any secret for polymorphic_url to work in the html view of the pdf?

**Comment by Samsinite** [comment]
For variants, I was able to get:
```erb
<%= image_tag Rails.application.routes.url_helpers.rails_representation_url(thing.attachment.variant(resize_to_limit: [300, 300])) %>
```

to work, you'll need to make sure that `default_url_options[:host]` is set for any `_url` helper to work. Using just `rails_representation_url` did not work, and needed to use `Rails.application.routes.url_helpers.rails_representation_url`.

**Comment by deepakmahakale** [comment]
None of the above solutions are working anymore on rails 6 or 7 

- Jun 2022

**Comment by supairish** [comment]
@deepakmahakale  I'm using macfanatic's helper with Rails 7 and it works fine, other than Rails 7 deprecating "service_url" for just "url".

**Comment by deepakmahakale** [comment]
Yes, I did those changes I also had to use strings for comparison as the code was raising [Uninitialized constant ActiveStorage::Service::DiskService](https://discuss.rubyonrails.org/t/uninitialized-constant-activestorage-diskservice/79226)

```ruby
  def wicked_blob_path(active_storage_attachment)
    service = active_storage_attachment.service
    service_name = service.class.name

    case service_name
    when 'ActiveStorage::Service::DiskService'
      service.path_for(active_storage_attachment.blob.key)
    when 'ActiveStorage::Service::S3Service'
      active_storage_attachment.url
    else
      raise 'Unsupported ActiveStorage service for WickedPDF integration: '\
            "#{service_name}"
    end
  end
```

This still doesn't work with `ActiveStorage::Service::DiskService`  in a rendered PDF.
